Develop

efrapp opened the pull request:
PR for code review.

orestis-kaplan committed:
First attempt of finishing the project

efrapp committed:
Adding validation to get the winner

efrapp committed:
Adding graphic table to show the game progress in the terminal

efrapp committed:
Adding table to guide the user what number match with the position of the table

efrapp committed:
Adding some methods private and putting some design to the GUI

efrapp committed:
Improving code and GUI

- Adding Draw case to the game
- Improving code in main file to use less lines of code
- Improving table "desing" for table guide and table game
- Adding "Do you want to play again?" functionality
- Avoid to use a position already taken

efrapp committed:
Adding default values

- Default value for players' name
- Default value for quit the game

efrapp committed:
Update README.md

efrapp committed:
Adding support for clear screen command on windows

efrapp committed:
Merge branch 'develop' of github.com:userman95/tic-tac-toe into develop

heyapricot commented on board.rb:
```diff
@@ -0,0 +1,73 @@
+    @size = size < 3 ? DIMENSION : size
+    @table = Array.new(@size)
+    @size.times { |pos| @table[pos] = Array.new(@size) }
+  end
```
What if you set size as a parameter with a default value?
`def initialize(size = 3)`

heyapricot replied:
Even better, why not ask for `width` and `height`?

efrapp replied:
@heyapricot in the case of the default value for the board, the `gets` method will never return `nil`, at least we will get an empty string or `0` when we are using the `to_i` method.

heyapricot commented on board.rb:
```diff
@@ -0,0 +1,73 @@
+    end
+
+    element
+  end
```
- Try using [string interpolation](http://ruby-for-beginners.rubymonstas.org/bonus/string_interpolation.html).
- Also, try generating a sequence of rows and columns with `Range` and `to_a`
- Favor the use of `each` over `for`

heyapricot replied:
Actually, with the way your `@table` is defined you can do this very easy with `each` and string interpolation

heyapricot commented on board.rb:
```diff
@@ -0,0 +1,73 @@
+
+  DIMENSION = 3
+
+  attr_accessor :size, :table
```
If `:table` is open to writing, somebody could alter the array directly. What would the implications be for that?

heyapricot commented on board.rb:
```diff
@@ -0,0 +1,73 @@
+  end
+
+  def empty_position?(arr_pos)
+    @table[arr_pos[0]][arr_pos[1]] == nil
```
Favor the use of `.nil?`

heyapricot commented on board.rb:
```diff
@@ -0,0 +1,73 @@
+    else
+      return false
+    end
+  end
```
Here's an idea:
`@table[arr_pos[0]][arr_pos[1]]= symbol if empty_position? arr_pos`
`cell == symbol`

Even better, what if you define a method to get the content of a cell and one for writing a cell?
`write_cell(arr_pos, symbol) if empty_position? arr_pos`
`read_cell(arr_pos) == symbol`

heyapricot commented on board.rb:
```diff
@@ -0,0 +1,73 @@
+  def table_guide
+    table_str = ""
+    (1..@size**2).each do |num|
+      table_str += "#{num}" + ' ' * ((@size * @size).to_s.length - num.to_s.length) + '|'
```
Why did you use concatenation and string interpolation? Why not string interpolation only?:
`"#{num}#{' ' * ((@size * @size).to_s.length - num.to_s.length)}|"`

heyapricot commented on board.rb:
```diff
@@ -0,0 +1,73 @@
+      end
+    end
+    table_str
+   end
```
This could be simplified a lot with string interpolaton

heyapricot commented on player.rb:
```diff
@@ -0,0 +1,19 @@
+  PLAYER1_SYMBOL = 'X'
+  PLAYER1_NAME = 'Player 1'
+  PLAYER2_SYMBOL = 'O'
+  PLAYER2_NAME = 'Player 2'
```
Why constants? These should be parameters

efrapp replied:
The names are wrong, they are just default values in case the user doesn't type one. We renamed them for better understanding, for example, `PLAYER1_NAME` is now `PLAYER1_DEFAULT_NAME`.

heyapricot commented on game.rb:
```diff
@@ -0,0 +1,100 @@
+      elsif i.all? { |inner| inner == 'O' }
+        return WIN
+      end
+    end
```
[There are easier ways to check if all the elements of the array are equal](https://stackoverflow.com/questions/27471254/ruby-check-if-all-array-elements-are-equal).
You just need to check if all the elements are equal, not if they are equal to a particular symbol

heyapricot replied:
Actually, because of this check if you use a custom symbol the program won't work

heyapricot commented on game.rb:
```diff
@@ -0,0 +1,100 @@
+      end
+      column_matrix << @columns
+      @columns = []
+    end
```
Here's an idea:
What if you use `map` to generate an array of arrays that arranges the contents of the column in the same fashion as the rows?
Something like:
`[[1, 4, 7], [2,5,8], [3,6,9] ]`

Then you could use the same method that checks that all the elements are the same :D

You could also create a method that returns the diagonal values as arrays 😏

heyapricot reviewed (changes requested):
# Good first attempt. There is room for improvement 🧼
- Game
1. The methods that check if a sequence has been completed belong to the board. Don't you think?
2. Game should take care of the flow of the game, answer questions like `who's turn is it` and `is the game over yet?`, also `who won?`
3. Game should also manage the Players (Instantiate them, assign the symbol)

- Board
1. Board should be able to answer `is there a sequence` and `is the board full?` as well as `show the current state of the board in the terminal`

Please address the comments and after that submit a new code review request. Thanks! 🤖

efrapp committed:
Applying code review

efrapp committed:
Organizing the game logic applying code review suggestion

efrapp committed:
Applying code reviewer suggestion

heyapricot commented on board.rb:
```diff
@@ -15,45 +18,93 @@ def graphic_table
     table_str = ""
     (1..@size**2).each do |num|
-      table_str += "#{num}" + ' ' * ((@size * @size).to_s.length - num.to_s.length) + '|'
+      table_str += "#{num}#{ ' ' * ((@size * @size).to_s.length - num.to_s.length) }|"
```
The second block on the string interpolation is doing nothing. `{ ' ' * ((@size * @size).to_s.length - num.to_s.length) }`
Let's think about it: `@size * @ size` in the regular case is going to equal `9`
`9.to_s` results in `'9'`
`'9'.length` equals `1`
`num` will have values from `1` to `9`
It is only one character so the length will be `1`
`1 - 1` equals `0`
`anything` times `0` equals `0`
The `' '` is being totally ignored

efrapp replied:
Hi @heyapricot the purpose of this code is that the table guide always looks square. For a 3x3 matrix you won't see the effect because we have only one digit on each cell but when the matrix is bigger (for example, two digits in some cells) we want that the cells with just one digit have the same space as the cell with two digits.

heyapricot replied:
I understand, but do you see the problem of calling `to_s` and then `length`? It returns the length of the number converted to a string. E.g. `100 * 100` = `10000` `10000.to_s` = `'10000'` `'10000'.length` = `5`

Is that what you are trying to achieve? I don't think so

efrapp replied:
That is what I'm trying to achieve because I need to count how many digits the number has to know how many spaces I'm going to put there and if I use the integer instead of the string (`1000.length`) I will get an error because `length` is not defined for numeric values.

heyapricot commented on board.rb:
```diff
@@ -15,45 +18,93 @@ def graphic_table
       if(num % @size).zero?
-        table_str += "\n" + '-' * ((@size * @size).to_s.length * @size + @size) + "\n"
+        table_str += "\n"
+        table_str += "#{'-' * ((@size * @size).to_s.length * @size + @size)} \n"
```
Again, the second part of the block multiplies by `1` so it's doing nothing. 
`@size + @size` is the same as `2 * @size`
This can be done in one line and also without the use of the `+` sign for string concatenation

efrapp replied:
Here we are multiplying the left `@size` with the `(@size * @size).to_s.length` first and the result of that with the right `@size` if we use `2 * @size` we will get a different result.
At the end we are trying to do this:
`(((@size * @size).to_s.length * @size) + @size)` but taking advantage of the operator's precedence, we omit the parenthesis.

heyapricot commented on board.rb:
```diff
@@ -15,45 +18,93 @@ def graphic_table
+   def check_rows
+    for row in @table
+      return WIN if row.uniq.size <= 1 && row.uniq != [nil]
+    end
```
Favor the use of `each`. Also, this could be one line

efrapp replied:
Applied.

heyapricot commented on board.rb:
```diff
@@ -15,45 +18,93 @@ def graphic_table
+
+    ON_GAME
+  end
+
```
Lines 64, 76 and 88 are almost identical. Also after every one of those lines ther is a return of the same constant. Surely there is a way to make a general case for these. A method that runs all the checks maybe? 🤔

efrapp replied:
@heyapricot about to method to run all the checks we think this is a responsibility of the game and we already have a method there (the `status` method) to check that.
About the line that is identical for all the check, we consider that we need it there because each method checks if there is win combination.

heyapricot replied:
Let's take a look at the `status` method inside `Game`
It calls the 4 methods that check the possible combinations. 
What I'm trying to say is that `Game` shouldn't even know or care about how the board determines the win (Calling each of `Board` methods for combination check)

Wouldn't it be nice to call something like `board.has_sequence?` and just get `true` or `false`?  

This `board.has_sequence?` can perform all the checks. `Game` only cares if there is a sequence, not which one

Let me know what you think. We can talk over here or on Crisp

heyapricot commented on board.rb:
```diff
@@ -15,45 +18,93 @@ def graphic_table
+  end
+
+  def check_main_diagonal
+    main_diagonal = []
```
This is reassigned in the next line. You can safely erase this line

heyapricot commented on board.rb:
```diff
@@ -15,45 +18,93 @@ def graphic_table
+  end
+
+  def check_second_diagonal
+    second_diagonal = []
```
This is reassigned in the next line. You can safely erase this line

heyapricot commented on board.rb:
```diff
@@ -15,45 +18,93 @@ def graphic_table
+    else
+      return false
+    end
+  end
```
Isn't this method redundant? Why not simply call `empty_position`?

efrapp replied:
Resolved.

heyapricot commented on board.rb:
```diff
@@ -67,7 +118,14 @@ def position_coverter position
     [row, position - @size * row - 1]
   end
```
You never use `col`. 
Favor the use of `each`. Also, this could be one line

efrapp replied:
Resolved.

heyapricot commented on game.rb:
```diff
@@ -1,100 +1,43 @@
+    draw_status = @board.draw
+    return draw_status if draw_status == Board::DRAW
+    
+    
```
These statements are also very similar. How can repetition be avoided in this block? This comment is related to the use of a method that performs all the checks

efrapp replied:
Resolved.

heyapricot commented on game.rb:
```diff
@@ -1,100 +1,43 @@
+    @player1.play(move)
+    @board.fill_table(@player1.value, @player1.symbol)
+    @player1, @player2 = @player2, @player1
   end
```
I like when methods/objects do one thing and do it really well. This method can be misleading since it doesn't make a move only, it also switches players and displays the board content in the console. If I wanted to build a tic tac toe using your class for the backend and called the make_a_move method I'll be displaying the board contents in the console and switching players indirectly. Should the make_a_move method take care of those? I think they should be separate methods.

efrapp replied:
Thanks Alvaro. Changed.

heyapricot commented on main.rb:
```diff
@@ -5,48 +5,45 @@
     
-    player_name, player_sym = setup_player(2, Player::PLAYER2_NAME, Player::PLAYER2_SYMBOL)
-    player2 = Player.new(player_name, player_sym)
+    player_name2 = setup_player(2, Player::PLAYER2_DEFAULT_NAME)
```
Is it really `main`'s responsibility to  instantiate the players? I think this should be the 'simplest' file of them all

efrapp replied:
@heyapricot we already move the logic for the game class, the `setup_player` is just to get the information from the console and validate the empty case.

heyapricot replied:
This is tricky. `Player` is not required in `Game`, yet `Game` is *aware* that a class named `Player` exists and that it has constants. 

It'd be simpler/easier if `main` just takes care of asking for a string to assign to the player and that's it. I don't think `main` should take care of what to do if the string is empty. That is `Player`'s responsibility in my opinion.

heyapricot commented on main.rb:
```diff
@@ -95,4 +90,4 @@ def print_table(board)
   puts board.graphic_table
 end
```
I think this is `Game`'s responsibility. Don't you think?

efrapp replied:
Changed.

heyapricot commented on main.rb:
```diff
@@ -5,48 +5,45 @@
 
```
Main should only require `game`

orestis-kaplan replied:
Resolved.

heyapricot commented on main.rb:
```diff
@@ -59,31 +56,29 @@ def welcome_screen
 
-def make_a_move(player, board)
+def setup_move(player, board)
   begin
```
[The `begin` here is redundant](https://www.rubydoc.info/gems/rubocop/0.27.0/RuboCop/Cop/Style/RedundantBegin)

orestis-kaplan replied:
Deleted it

heyapricot commented on main.rb:
```diff
@@ -59,31 +56,29 @@ def welcome_screen
     puts "Enter the value from 1 to #{board.size**2}"
     print "#{player.name} your turn: "
     next_move = gets.match(/\d+/)[0].to_i
     raise "Please enter a number in the range." unless (next_move).between?(1, board.size**2)
```
Redundant parentheses on `next_move`

orestis-kaplan replied:
Resolved.

heyapricot commented on main.rb:
```diff
@@ -59,31 +56,29 @@ def welcome_screen
     puts "Please enter only numeric values"
     retry
   rescue RuntimeError => e
     puts "#{e}"
```
Prefer `to_s` over string interpolation

orestis-kaplan replied:
@heyapricot  In which line should we  put the to_s? if we place it in ` puts "Enter the value from 1 to #{board.size**2}"` the solution would be `puts "Enter the value from 1 to"+(board.size**2).to_s` 
but why not use an interpolation here ?

heyapricot replied:
Yes, this one is on me. I wasn't clear enough 🤦🏻‍♂️
When you need the result string to contain *ONLY* the content of a variable, use `to_s`
Otherwise, use string interpolation (E.g if you need to put the contents of a variable in a message string)

Is this clear?

efrapp replied:
Ok got it and changed.

heyapricot commented on main.rb:
```diff
@@ -59,31 +56,29 @@ def welcome_screen
   welcome = "*"*welcome.length + "\n" + welcome + "\n" + "*"*welcome.length
 end
```
This is very tricky. Why not just one line with string interpolation?

orestis-kaplan replied:
Resolved

heyapricot commented on main.rb:
```diff
@@ -59,31 +56,29 @@ def welcome_screen
+  else
+    return next_move
   end
 end
```
A pretty cool job using `rescue` and handling exceptions 😎

orestis-kaplan replied:
nice 👍

heyapricot reviewed (changes requested):
# Less is more 🧶
Hello! I've added comments to improve the project. One of the things I focused on was finding ways to avoid repetitions. I also left comments for encouraging solutions on `one line`. These are not strictly required but I think they can look beautiful and make the code simpler to understand. 

Please, address the coments I've left and after that submit a new code review request. 

Feel free to reach me over crisp to discuss any comment I've left. Thank you!

efrapp committed:
Applying second code review changes

orestis-kaplan committed:
Resolving issues

orestis-kaplan committed:
deleting extra files

heyapricot commented on main.rb:
```diff
@@ -53,24 +52,23 @@ def main
   print "Please insert the name of the player #{num_of_player}: "
   player_name = gets.strip.capitalize
-  player_name = player_name == "" ? name : player_name
+  player_name = player_name == "" ? default_name : player_name
```
Is it `main`'s responsibility to sanitize this input? Is it really necessary? I think not.
Sanitizing is not bad per se but if you make `Player` protect against an empty name you can omit this.

If for some reason you still want to sanitize you could do something like:
`player_name = default_name if player_name.length < 1` which is easier to read/understand

orestis-kaplan replied:
Resolved

heyapricot commented on main.rb:
```diff
@@ -38,7 +37,7 @@ def main
       else
-        game.change_player
+        game.switch_players
       end
```
Same as Player constants. even though the `Board` class is not directly required in your `main` method, `main` is aware that a class named `Board` exists and that it has `constants`. This makes `main` susceptible to being broken if `board` is modified for any reason. 

We can break this problem into sub problems.
1. You want to know if the game is still active. You could know that in main by asking `Game`. Something like `Game.still_active?`
2. You want to know how the game ended so you put the appropiate message. They way I see this is:
 - You ask if the game is still active
- If the game is not active anymore that means either there is a sequence or the board is full. How can `game ` return that information to `main` so that `main` can take the appropiate decision (`main` asks: what do I print?)

heyapricot commented on game.rb:
```diff
@@ -13,31 +15,42 @@ def initialize(board_dimension, player1_name, player1_sym, player2_name, player2
-    @board.fill_table(@player1.value, @player1.symbol)
-    @player1, @player2 = @player2, @player1
+    @player_on_turn.play(move)
+  end
```
How does your scoring work? If I choose cell 9 I get 9 points?

orestis-kaplan replied:
Currently, we don't have a scoring system . The cell that is being chosen ,it appears in the board and the check for win status is in process

heyapricot reviewed (changes requested):
# Let's keep thing isolated 🧫
The game is looking really good. One thing I've notice in this review is that there still is a non-explicit dependancy between classes. What we want to achieve is full isolation. Why? Well, currently the classes are dependant of knowing the contents of the other classes (Particularly with constants) if at some point we had to change the constants (Or any other thing) this would break our classes and debugging would be hard becuase we would say: *Hey! Main was only dependant of game. What happened?*

This is a very important topic to be aware of. I thank you for your patince and following the reviews this far

orestis-kaplan committed:
validation the player class

orestis-kaplan committed:
Resolving remaining issues

orestis-kaplan committed:
Resolving remaining issues

heyapricot commented on game.rb:
```diff
@@ -13,23 +13,24 @@ def initialize(board_dimension, player1_name, player1_sym, player2_name, player2
+      return false
     end
-    Board::ON_GAME
+  end
```
Nice! This can be a beautiful one-liner:
`def still_active? (status)`
`status == Board::ON_GAME`
`end`

heyapricot commented on main.rb:
```diff
@@ -8,11 +8,11 @@ def main
+    game = Game.new(dimension, player_name1, player1_symbol, player_name2, player2_symbol)
 
     on_game = true
     play_again = true
```
You assign this variable again on line 46 and never use it before that so what's the point of initializing it?

heyapricot commented on main.rb:
```diff
@@ -8,11 +8,11 @@ def main
-    game = Game.new(dimension, player_name1, Player::PLAYER1_DEFAULT_SYMBOL, player_name2, Player::PLAYER2_DEFAULT_SYMBOL)
+    game = Game.new(dimension, player_name1, player1_symbol, player_name2, player2_symbol)
 
     on_game = true
```
You are ignoring the assignment of this variable in line 47. If you repeat the loop it will be true again. (It doesn't matter which value you assign on line 47, when the loop is repeated the variable will be set to true)

heyapricot reviewed (changes requested):
# Let's make a wish 🧞‍♀️
This is how I think the main loop could look like. I tried not to differ a lot from the current implementation. Let me know what you think and if you agree, let's do it 

```def main
  loop do
    welcome_screen
    game = Game.new(get_dimension)
    print_board
    while game.still_active?
      game.make_a_move(setup_move)
      print_board
      game.switch_players
    end
    # At this point the game is not active anymore
    announce_winner
    break unless play_again?
  end
  goodbye
end
```

You already have all these elements but they are scattered around. Try building the missing methods
and improving the ones you already have along with your classes. Also be mindful of which method goes where.

Missing: `print_board`, `still_active?`, `announce_winner`, `play_again?`, `goodbye`

orestis-kaplan committed:
adding functions to the main and take care of the loop

efrapp committed:
Creating set_player method for Game class

heyapricot commented on main.rb:
```diff
@@ -6,57 +6,60 @@ def main
     puts welcome_screen
```
Let's get the `puts` inside the welcome_screen method

efrapp replied:
Done.

heyapricot commented on main.rb:
```diff
@@ -6,57 +6,60 @@ def main
-    dimension = gets.strip.to_i
 
-    player_name1,player1_symbol = setup_player(1)
+    dimension = gets.strip.to_i
```
Can we store these two in a method? Something like `get_dimension`

efrapp replied:
Alvaro when you say "these two", do you mean `dimension = gets.strip.to_i` and this `game = Game.new(dimension)` ? if so, maybe we can leave the initialization of the game outside of that method (I mean, in the place where it is now) because the method should to return the game and I don't see the point of that.

heyapricot replied:
It's good in its current state

efrapp replied:
👍

heyapricot commented on main.rb:
```diff
@@ -6,57 +6,60 @@ def main
-    on_game = true
-    play_again = true
+    player_name2,player2_symbol = get_player_info(2)
+    game.set_player(1, player_name2, player2_symbol)
```
Lines 14 to 18 are almost identical. The only thing that is changing is the player whose attributes you are setting up. Can a method be defined for this? The input could be the player you want to modify and the information you want to set for the player

efrapp replied:
@heyapricot if we put those lines in one method, we think that the method will be called twice as we make now to set up the two players, so what is the point of that? maybe reduce the numbers of lines?

heyapricot replied:
This makes the project error prone. By defining a method we know for sure you are doing EXACTLY the same actions for both users. Whereas if you copy-paste there is a chance you get a typo. Or maybe you put a line in between the actions of user one that you don't for user two

efrapp replied:
Ok nice. We will update this. Now I understand the purpose. Thank you.

heyapricot commented on game.rb:
```diff
@@ -3,25 +3,22 @@
-      return false
-    end
+     status == Board::ON_GAME
   end
```
The current implementation for the board status is confusing. You have constants for `WIN` and `DRAW` on board.rb. The `return_status` method is confusing since it does multiple things: It checks if the game was won or if it's a draw and then returns the status. 

My suggestions:
1. Choose which class is going to take full responsibility of deciding if the game has a winner or if it's a draw (I suggest `Game`
2. Have two methods. One that tells if the board is full or not and another to determine the winner

heyapricot replied:
Also, I don't think any of these methods need a status as input

orestis-kaplan replied:
Removed the status as input

heyapricot commented on main.rb:
```diff
@@ -6,57 +6,60 @@ def main
+  next_move = setup_move(game.player_on_turn, game.board)
+  game.make_a_move(next_move)
+  game.fill_table
+end
```
You mentioned you are trying to follow the [Facade Pattern ](https://en.wikipedia.org/wiki/Facade_pattern) so let's remove the object's dependencies between each other. This should look like:
```
  next_move = setup_move(current_player, board)
  game.make_a_move(next_move)
  game.fill_table
```

efrapp replied:
@heyapricot what would be `current_player`? is it a variable with `game.player_on_turn` assigned on it or a function (in the same `main` file) that calls `game.player_on_turn`? I think the object's dependencies you mention are not related to the Facade Pattern, it is related about to make a clear separation between the core application and the interface we are using as a source of input (in our case the `main` file). If we use another facade (for the web, for example) the code will be almost the same we have for the `main` file just with a different source of input (text fields and some things like that) but things like `game.player_on_turn` won't break the pattern 
because that doesn't affect the separation of input sources.

heyapricot replied:
Alright, we can omit this.

efrapp replied:
👍

heyapricot commented on main.rb:
```diff
@@ -6,57 +6,60 @@ def main
-        next_move = setup_move(game.player_on_turn, game.board)
-        game.make_a_move(next_move)
-        game.fill_table
+        print_table(game)
```
You can add the `system("cls") || system("clear")` to the print_table method
Also, I think the way to call this method should be `game.print_table` or `game.print_guide`

heyapricot reviewed (changes requested):
# Can Game be the Facade🚪?

As `main` becomes more complex and `game` becomes simpler I wonder, can `game` work as the facade that handles `board` and `player` ? I say yes. Let's have a conversation about that on crsip 

- I left comments for some fixes. Please address those
- Rubocop doesn't like method definitions that don't have their parameters surrounded by parentheses. You have multiple occurrences of this. Please fix those.

We are almost there. Bear with me

efrapp committed:
Adding puts line inside the welcome screen method

orestis-kaplan committed:
adding suggestions for facede review

efrapp committed:
Improving setup players
- Using one method to setup both players
- Improving code to setup default values for both players
- Replacing require with require_relative

efrapp committed:
Solving merge conflict with remote develop

heyapricot reviewed (approved):
# It's been a long journey. Let's ship it 🚢
Alright team! You worked hard on this. Pat yourselves in the back. Hope that you were able to enjoy and learn from this procedure even if it wasn't super fun all times. No need from my side for further review
